lib/sisimai/bite/json/amazonses.rb

```diff
@@ -0,0 +1,255 @@
+      # @options mhead [String] subject   Subject header
+      # @options mhead [Array]  received  Received headers
+      # @options mhead [String] others    Other required headers
+      # @param         [String] mbody     Message body of a bounce email
+      # @return        [Hash, Nil]        Bounce data list and message/rfc822
+      #                                   part or nil if it failed to parse or
+      #                                   the arguments are missing
+      # @since v4.20.0
+      def scan(mhead, mbody)
+        return nil unless mhead
+        return nil unless mbody
+        return nil unless mhead['x-amz-sns-message-id']
+        return nil unless mhead['x-amz-sns-message-id'].size > 0
+
+        hasdivided = mbody.split("\n")
+        jsonstring = ''
+        jsonthings = nil
+        foldedline = false
+
+        hasdivided.each do |e|
+          # Find JSON string from the message body
+          next if e.size.zero?
+          break if e =~ /\A[-]{2}\z/
+          break if e == '__END_OF_EMAIL_MESSAGE__'
+
+          # The line starts with " ", continued from !\n.
+          e = e.sub(/\A[ ]/, '') if foldedline
+          foldedline = false
+
+          if e =~ /[!]\z/
+            # ... long long line ...![\n]
+            e = e.sub(/!\z/, '')
+            foldedline = true
+          end
+          jsonstring += e
+        end
+
+        begin
+          if RUBY_PLATFORM =~ /java/
+            # java-based ruby environment like JRuby.
+            require 'jrjackson'
+            jsonobject = JrJackson::Json.load(jsonstring)
+            if jsonobject['Message']
+              # 'Message' => '{"notificationType":"Bounce",...
+              jsonthings = JrJackson::Json.load(jsonobject['Message'])
+            end
+          else
+            # Matz' Ruby Implementation
+            require 'oj'
+            jsonobject = Oj.load(jsonstring)
+            if jsonobject['Message']
+              # 'Message' => '{"notificationType":"Bounce",...
+              jsonthings = Oj.load(jsonobject['Message'])
+            end
+          end
+          jsonthings ||= jsonobject
+
+        rescue StandardError => ce
+          # Something wrong in decoding JSON
+          warn sprintf(' ***warning: Failed to decode JSON: %s', ce.to_s)
+          return nil
+        end
+
+        return adapt(jsonthings)
+      end
+
+      # @abstract Adapt Amazon SES bounce object for Sisimai::Message format
+      # @param        [Hash] argvs  bounce object(JSON) retrieved from Amazon SNS
+      # @return       [Hash, Nil]   Bounce data list and message/rfc822 part
+      #                             or Undef if it failed to parse or the
+      #                             arguments are missing
+      # @since v4.20.0
+      def adapt(argvs)
+        return nil unless argvs.is_a? Hash
+        return nil unless argvs.keys.size > 0
+        return nil unless argvs.key?('notificationType')
+
```

Comment:
__[rubocop]__
* Metrics/AbcSize: Assignment Branch Condition size for adapt is too high. [112.8/30] [:ref](http://c2.com/cgi/wiki?AbcMetric)

<!-- Comment by [SideCI](https://sideci.com) -->